File: backend/app/api/v1/endpoints/match.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.schemas import (
    BatchMatchRequest,
    BatchTaskResponse,
    BatchTaskStatusResponse,
    MatchScoreRequest,
    MatchScoreResponse,
)
from app.services.match import (
    MatchNotFoundError,
    MatchService,
    MatchSkillError,
    ResumeNotParsedError,
)

router = APIRouter(prefix="/match-scores", tags=["人岗匹配"])


@router.post("", response_model=MatchScoreResponse, status_code=201, summary="单点触发人岗匹配")
async def create_match_score(request: MatchScoreRequest, db: AsyncSession = Depends(get_db)):
    service = MatchService(db)
    try:
        score = await service.match_one(request.jd_id, request.resume_id, force=request.force)
    except ResumeNotParsedError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except MatchNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except MatchSkillError as e:
        raise HTTPException(status_code=500, detail=str(e))
    resp = MatchScoreResponse.model_validate(score)
    resp.is_stale = await service.compute_is_stale(score)
    return resp


@router.post("/batch", response_model=BatchTaskResponse, status_code=202, summary="批量触发人岗匹配")
async def create_batch_match(request: BatchMatchRequest, db: AsyncSession = Depends(get_db)):
    service = MatchService(db)
    try:
        handle = await service.batch_match(
            request.jd_id,
            resume_ids=request.resume_ids,
            limit=request.limit,
            force=request.force,
        )
    except MatchNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    return BatchTaskResponse(
        task_id=handle.task_id,
        jd_id=handle.jd_id,
        total_submitted=handle.total_submitted,
        submitted_at=handle.submitted_at,
    )


@router.get("/batch/{task_id}", response_model=BatchTaskStatusResponse, summary="查询批量匹配任务状态")
async def get_batch_status(task_id: str, db: AsyncSession = Depends(get_db)):
    service = MatchService(db)
    state = service.get_batch_status(task_id)
    if state is None:
        raise HTTPException(status_code=404, detail="batch task not found")
    return BatchTaskStatusResponse(
        task_id=state["task_id"],
        jd_id=state["jd_id"],
        total=state["total"],
        completed=state["completed"],
        failed=state["failed"],
        status=state["status"],
        started_at=state["started_at"],
        finished_at=state["finished_at"],
    )


@router.get("/{score_id}", response_model=MatchScoreResponse, summary="查询单条匹配详情")
async def get_match_score(score_id: str, db: AsyncSession = Depends(get_db)):
    service = MatchService(db)
    score = await service.get_score(score_id)
    if score is None:
        raise HTTPException(status_code=404, detail="match score not found")
    resp = MatchScoreResponse.model_validate(score)
    resp.is_stale = await service.compute_is_stale(score)
    return resp

File: frontend/src/services/match.ts
import request from '@/utils/request';
import type {
  MatchScore,
  MatchRankingResponse,
  BatchTaskResponse,
  BatchTaskStatus,
} from '@/types';

export interface MatchOneRequest {
  jd_id: string;
  resume_id: string;
  force?: boolean;
}

export interface BatchMatchRequest {
  jd_id: string;
  resume_ids?: string[];
  limit?: number;
  force?: boolean;
}

export const matchApi = {
  matchOne: (data: MatchOneRequest) =>
    request.post<any, MatchScore>('/match-scores', data),

  batchMatch: (data: BatchMatchRequest) =>
    request.post<any, BatchTaskResponse>('/match-scores/batch', data),

  getBatchStatus: (taskId: string) =>
    request.get<any, BatchTaskStatus>(`/match-scores/batch/${taskId}`),

  getScore: (scoreId: string) =>
    request.get<any, MatchScore>(`/match-scores/${scoreId}`),

  rankByJd: (jdId: string, params?: { limit?: number; offset?: number }) =>
    request.get<any, MatchRankingResponse>(`/jds/${jdId}/ranking`, { params }),

  listByResume: (resumeId: string, params?: { limit?: number }) =>
    request.get<any, MatchScore[]>(`/resumes/${resumeId}/matches`, { params }),
};
